File: 1.Two_Sum/sol.py
import os
import sys
from typing import List

# ======================================================================
# 🧠 CHALLENGE: Two Sum (Python Version)
# ======================================================================
# Description:
# Given an array of integers `nums` and an integer `target`, return
# indices of the two numbers such that they add up to `target`.
#
# 📋 Rules:
# 1. You may assume that each input would have EXACTLY one solution.
# 2. You may not use the same element twice.
# 3. You can return the answer in any order.
# 4. Follow-up: Can you come up with an algorithm that is less than
#    O(n^2) time complexity?
#
# 💡 Examples:
# - two_sum([2,7,11,15], 9) => [0, 1]
# - two_sum([3,2,4], 6)     => [1, 2]
# - two_sum([3,3], 6)       => [0, 1]
# ======================================================================

# #region [📚 Reference Solutions]


def two_sum1(nums: List[int], target: int) -> List[int]:
    """
    Method: One-pass Hash Table
    Complexity: Time O(N) | Space O(N)
    """
    # Dictionary to store number: index
    seen = {}

    for i, num in enumerate(nums):
        complement = target - num

        # Check if the complement exists in our dictionary
        if complement in seen:
            return [seen[complement], i]

        # Store current number and index
        seen[num] = i

    return []  # Should not happen per problem description


# #endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def two_sum(nums, target):
    # TODO: Implement your solution here.
    return []


# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def two_sum(nums, target):\n",
        "    # TODO: Implement your solution here.\n",
        "    return []\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)
        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ([2, 7, 11, 15], 9, [0, 1]),
        ([3, 2, 4], 6, [1, 2]),
        ([3, 3], 6, [0, 1]),
    ]

    print(f"\n🧪 Testing your [two_sum] function...\n")

    header = f"{'Input nums':<20} | {'Target':<6} | {'Expected':<10} | {'Actual':<10} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for nums, target, expected in test_cases:
        try:
            result = two_sum(nums, target)
        except Exception as e:
            result = "Error"

        # Compare sorted lists to allow returning indices in any order
        is_match = False
        if isinstance(result, list) and len(result) == 2:
            is_match = sorted(result) == sorted(expected)

        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        # Format output
        nums_str = str(nums)
        if len(nums_str) > 18:
            nums_str = nums_str[:15] + "..."

        print(
            f"{nums_str:<20} | {str(target):<6} | {str(expected):<10} | {str(result):<10} | {status_icon}"
        )

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")

# endregion

File: 1.Two_Sum/sol.go
package main

import (
	"fmt"
	"os"
	"reflect"
	"runtime"
	"sort"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Two Sum (Go Version)
// ======================================================================
// Description:
// Given an array of integers `nums` and an integer `target`, return
// indices of the two numbers such that they add up to `target`.
//
// 📋 Rules:
// 1. You may assume that each input would have EXACTLY one solution.
// 2. You may not use the same element twice.
// 3. You can return the answer in any order.
// 4. Follow-up: Can you come up with an algorithm that is less than
//    O(n^2) time complexity?
//
// 💡 Examples:
// - TwoSum([2,7,11,15], 9) => [0, 1]
// - TwoSum([3,2,4], 6)     => [1, 2]
// - TwoSum([3,3], 6)       => [0, 1]
// ======================================================================

// #region [📚 Reference Solutions]

/**
 * Method: One-pass Hash Map
 * Complexity: Time O(N) | Space O(N)
 */
func TwoSum1(nums []int, target int) []int {
	// Create a map to store value -> index
	seen := make(map[int]int)

	for i, num := range nums {
		complement := target - num

		// Check if complement exists in map
		if idx, ok := seen[complement]; ok {
			return []int{idx, i}
		}

		// Store current number and index
		seen[num] = i
	}

	return nil // Should not be reached
}

// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func TwoSum(nums []int, target int) []int {
	// TODO: Implement your solution here.
	return []int{}
}

// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

func main() {
	runTests()
}

type TestCase struct {
	nums     []int
	target   int
	expected []int
}

func runTests() {
	testCases := []TestCase{
		{[]int{2, 7, 11, 15}, 9, []int{0, 1}},
		{[]int{3, 2, 4}, 6, []int{1, 2}},
		{[]int{3, 3}, 6, []int{0, 1}},
	}

	fmt.Printf("\n🧪 Testing your [TwoSum] function...\n\n")

	header := fmt.Sprintf("%-20s | %-6s | %-10s | %-10s | Status", "Input nums", "Target", "Expected", "Actual")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		result := TwoSum(tc.nums, tc.target)

		// Create copies to sort for comparison without modifying originals (though these are local)
		resCopy := make([]int, len(result))
		copy(resCopy, result)
		expCopy := make([]int, len(tc.expected))
		copy(expCopy, tc.expected)

		sort.Ints(resCopy)
		sort.Ints(expCopy)

		isMatch := reflect.DeepEqual(resCopy, expCopy)
		statusIcon := "✅ PASS"
		if !isMatch {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		// Helper to format slices
		fmtSlice := func(s []int) string {
			return fmt.Sprintf("%v", s)
		}

		numsStr := fmtSlice(tc.nums)
		if len(numsStr) > 18 {
			numsStr = numsStr[:15] + "..."
		}

		fmt.Printf("%-20s | %-6d | %-10s | %-10s | %s\n",
			numsStr, tc.target, fmtSlice(tc.expected), fmtSlice(result), statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func TwoSum(nums []int, target int) []int {",
		"\t// TODO: Implement your solution here.",
		"\treturn []int{}",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion
